cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -93,53 +123,212 @@ std::unique_ptr<column> rank_generator(column_view const& order_by,
+      size_type row_index) {
+      auto group_start = offsets[labels[row_index]];
+      auto group_end   = offsets[labels[row_index] + 1];
+      return resolver(row_index + 1 == group_end || !comparator(row_index, row_index + 1),
+                      row_index - group_start);
+    });
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                thrust::reverse_iterator(group_labels.end()),
+                                thrust::reverse_iterator(group_labels.begin()),
+                                thrust::reverse_iterator(mutable_ranks.end<size_type>()),
+                                thrust::reverse_iterator(mutable_ranks.end<size_type>()),
+                                thrust::equal_to{},
+                                scan_op);
+
+  return ranks;
+}
 }  // namespace
 
-std::unique_ptr<column> rank_scan(column_view const& order_by,
-                                  device_span<size_type const> group_labels,
-                                  device_span<size_type const> group_offsets,
-                                  rmm::cuda_stream_view stream,
-                                  rmm::mr::device_memory_resource* mr)
+std::unique_ptr<column> min_rank_scan(column_view const& grouped_values,
+                                      column_view const& value_order,
+                                      device_span<size_type const> group_labels,
+                                      device_span<size_type const> group_offsets,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
 {
   return rank_generator(
-    order_by,
+    grouped_values,
+    value_order,
     group_labels,
     group_offsets,
     [] __device__(bool unequal, auto row_index_in_group) {
       return unequal ? row_index_in_group + 1 : 0;
     },
     DeviceMax{},
-    has_nested_nulls(table_view{{order_by}}),
+    has_nested_nulls(table_view{{grouped_values}}),
     stream,
     mr);
 }
 
-std::unique_ptr<column> dense_rank_scan(column_view const& order_by,
+std::unique_ptr<column> max_rank_scan(column_view const& grouped_values,
+                                      column_view const& value_order,
+                                      device_span<size_type const> group_labels,
+                                      device_span<size_type const> group_offsets,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  return rank_generator_reverse(
+    grouped_values,
+    value_order,
+    group_labels,
+    group_offsets,
+    [] __device__(bool unequal, auto row_index_in_group) {
+      return unequal ? row_index_in_group + 1 : 0;
+    },
+    [] __device__(auto val1, auto val2) {
+      return val1 == 0 or val2 == 0 ? std::max(val1, val2) : std::min(val1, val2);
+    },
```

Comment:
I'm surprised that this scan op isn't just the inverse of the `DeviceMax` used above. I don't see any comparable checks for val1/val2 equaling 0 in the `min_rank_scan` implementation, should there be?

Reply:
My initial implementation used `numeric_limits<size_type>::max()` and `DeviceMin`. Since rank range is from 1 to `numeric_limits<size_type>::max()`, then `numeric_limits<size_type>::max() + 1` will cause overflow. Hence changed to this implementation.

Reply:
Hmm how does this get around that problem? Don't you always have an issue if the column contains the maximum number of elements and all of them are unique? Overflow is always possible for the final rank.

Reply:
Right. This code does not fix the problem I anticipated. 
Problem: (max no. of elements, all unique), For last index `numeric_limits<size_type>::max() + 1` will result in `numeric_limits<size_type>::min()`. Reverse iterator scan would result in all rank values to be `numeric_limits<size_type>::min()`.

Would this fix work?
```suggestion
    [] __device__(auto val1, auto val2) {
      return val1 <= 0 or val2 <= 0 ? std::max(val1, val2) : std::min(val1, val2);
    },
```

If you think, this is an insignificant corner case to be ignored, I prefer using `DeviceMin` and `numerical_limits<size_type>::max()`. What do you think?

Reply:
I agree, let's use just use `DeviceMin`. I think there are many algorithms in cudf that will fail in these kinds of cases where a column has 2^31-1 elements and one other condition is met (depending on the algorithm). If that condition is sufficiently rare I think we just live with it for now.